File: lib/core/network/supabase_data_proxy.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:marketflow/config/backend_proxy_config.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

class DataProxyFilter {
  final String column;
  final String op;
  final dynamic value;

  const DataProxyFilter(this.column, this.op, this.value);

  const DataProxyFilter.eq(String column, dynamic value)
    : this(column, 'eq', value);

  const DataProxyFilter.gt(String column, dynamic value)
    : this(column, 'gt', value);

  const DataProxyFilter.gte(String column, dynamic value)
    : this(column, 'gte', value);

  const DataProxyFilter.lt(String column, dynamic value)
    : this(column, 'lt', value);

  const DataProxyFilter.lte(String column, dynamic value)
    : this(column, 'lte', value);

  const DataProxyFilter.neq(String column, dynamic value)
    : this(column, 'neq', value);

  Map<String, dynamic> toJson() => {'column': column, 'op': op, 'value': value};
}

class DataProxyOrder {
  final String column;
  final bool ascending;
  final bool? nullsFirst;

  const DataProxyOrder(this.column, {this.ascending = true, this.nullsFirst});

  Map<String, dynamic> toJson() => {
    'column': column,
    'ascending': ascending,
    if (nullsFirst != null) 'nullsFirst': nullsFirst,
  };
}

class SupabaseDataProxy {
  final SupabaseClient _db;

  const SupabaseDataProxy({required SupabaseClient db}) : _db = db;

  Future<dynamic> rpc(
    String functionName, {
    Map<String, dynamic>? params,
  }) async {
    final normalizedName = functionName.trim();
    if (normalizedName.isEmpty) {
      throw const PostgrestException(
        message: 'RPC function name is required',
        code: 'INVALID_RPC_NAME',
      );
    }

    final proxyUri = _proxyUri();
    if (proxyUri == null) {
      return _db.rpc(normalizedName, params: params);
    }

    return _postProxy(proxyUri, <String, dynamic>{
      'kind': 'rpc',
      'rpc': normalizedName,
      'params': params ?? const <String, dynamic>{},
    });
  }

  Future<List<dynamic>> select({
    required String table,
    String columns = '*',
    List<DataProxyFilter> filters = const <DataProxyFilter>[],
    List<DataProxyOrder> orders = const <DataProxyOrder>[],
    int? limit,
  }) async {
    final normalizedTable = table.trim();
    if (normalizedTable.isEmpty) {
      throw const PostgrestException(
        message: 'Table name is required',
        code: 'INVALID_TABLE_NAME',
      );
    }

    final proxyUri = _proxyUri();
    if (proxyUri == null) {
      dynamic query = _db.from(normalizedTable).select(columns);
      query = _applyFilters(query, filters);
      query = _applyOrders(query, orders);
      if (limit != null) {
        query = query.limit(limit);
      }
      final result = await query;
      return _asList(result);
    }

    final result = await _postProxy(proxyUri, <String, dynamic>{
      'kind': 'table',
      'action': 'select',
      'table': normalizedTable,
      'columns': columns,
      'filters': filters.map((item) => item.toJson()).toList(),
      'orders': orders.map((item) => item.toJson()).toList(),
      ...?limit == null ? null : <String, dynamic>{'limit': limit},
    });
    return _asList(result);
  }

  Future<dynamic> update({
    required String table,
    required Map<String, dynamic> values,
    List<DataProxyFilter> filters = const <DataProxyFilter>[],
    bool returning = false,
    String columns = '*',
  }) async {
    final normalizedTable = table.trim();
    if (normalizedTable.isEmpty) {
      throw const PostgrestException(
        message: 'Table name is required',
        code: 'INVALID_TABLE_NAME',
      );
    }

    final proxyUri = _proxyUri();
    if (proxyUri == null) {
      dynamic query = _db.from(normalizedTable).update(values);
      query = _applyFilters(query, filters);
      if (returning) {
        query = query.select(columns);
      }
      return await query;
    }

    return _postProxy(proxyUri, <String, dynamic>{
      'kind': 'table',
      'action': 'update',
      'table': normalizedTable,
      'values': values,
      'filters': filters.map((item) => item.toJson()).toList(),
      'returning': returning,
      'columns': columns,
    });
  }

  Future<dynamic> upsert({
    required String table,
    required dynamic values,
    String? onConflict,
    bool returning = false,
    String columns = '*',
  }) async {
    final normalizedTable = table.trim();
    if (normalizedTable.isEmpty) {
      throw const PostgrestException(
        message: 'Table name is required',
        code: 'INVALID_TABLE_NAME',
      );
    }

    final proxyUri = _proxyUri();
    if (proxyUri == null) {
      dynamic query = _db
          .from(normalizedTable)
          .upsert(values, onConflict: onConflict);
      if (returning) {
        query = query.select(columns);
      }
      return await query;
    }

    return _postProxy(proxyUri, <String, dynamic>{
      'kind': 'table',
      'action': 'upsert',
      'table': normalizedTable,
      'values': values,
      if (onConflict != null && onConflict.trim().isNotEmpty)
        'onConflict': onConflict.trim(),
      'returning': returning,
      'columns': columns,
    });
  }

  dynamic _applyFilters(dynamic query, List<DataProxyFilter> filters) {
    dynamic current = query;
    for (final filter in filters) {
      final op = filter.op.trim().toLowerCase();
      switch (op) {
        case 'eq':
          current = current.eq(filter.column, filter.value);
          break;
        case 'neq':
          current = current.neq(filter.column, filter.value);
          break;
        case 'gt':
          current = current.gt(filter.column, filter.value);
          break;
        case 'gte':
          current = current.gte(filter.column, filter.value);
          break;
        case 'lt':
          current = current.lt(filter.column, filter.value);
          break;
        case 'lte':
          current = current.lte(filter.column, filter.value);
          break;
        default:
          throw PostgrestException(
            message: 'Unsupported filter operator: ${filter.op}',
            code: 'UNSUPPORTED_FILTER_OP',
          );
      }
    }
    return current;
  }

  dynamic _applyOrders(dynamic query, List<DataProxyOrder> orders) {
    dynamic current = query;
    for (final order in orders) {
      if (order.nullsFirst == null) {
        current = current.order(order.column, ascending: order.ascending);
      } else {
        current = current.order(
          order.column,
          ascending: order.ascending,
          nullsFirst: order.nullsFirst,
        );
      }
    }
    return current;
  }

  Future<dynamic> _postProxy(Uri proxyUri, Map<String, dynamic> payload) async {
    final headers = <String, String>{'Content-Type': 'application/json'};
    final token = _db.auth.currentSession?.accessToken.trim() ?? '';
    if (token.isNotEmpty) {
      headers['Authorization'] = 'Bearer $token';
    }

    http.Response response;
    try {
      response = await http
          .post(proxyUri, headers: headers, body: jsonEncode(payload))
          .timeout(Duration(seconds: BackendProxyConfig.timeoutSeconds));
    } on TimeoutException {
      throw const PostgrestException(
        message: 'Backend data proxy timed out',
        code: 'PROXY_TIMEOUT',
      );
    } catch (error) {
      throw PostgrestException(
        message: error.toString(),
        code: 'PROXY_REQUEST_FAILED',
      );
    }

    final decoded = _decodeResponse(response.body);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      if (decoded is Map<String, dynamic>) {
        throw PostgrestException(
          message: (decoded['message'] ?? 'Database request failed').toString(),
          code: decoded['code']?.toString(),
          details: decoded['details'],
          hint: decoded['hint']?.toString(),
        );
      }
      throw PostgrestException(
        message: response.body.trim().isEmpty
            ? 'Database request failed'
            : response.body.trim(),
        code: 'PROXY_HTTP_${response.statusCode}',
      );
    }

    return decoded;
  }

  dynamic _decodeResponse(String raw) {
    final value = raw.trim();
    if (value.isEmpty) return const <String, dynamic>{};
    try {
      return jsonDecode(value);
    } catch (_) {
      return value;
    }
  }

  List<dynamic> _asList(dynamic raw) {
    if (raw is List<dynamic>) return raw;
    if (raw == null) return const <dynamic>[];
    if (raw is List) return List<dynamic>.from(raw);
    return <dynamic>[raw];
  }

  Uri? _proxyUri() {
    if (!BackendProxyConfig.isDataProxyEnabled) return null;
    final raw = BackendProxyConfig.dataEndpoint.trim();
    if (raw.isEmpty) return null;

    final parsed = Uri.tryParse(raw);
    if (parsed == null) return null;
    if (!parsed.hasScheme && !kIsWeb) return null;
    return parsed;
  }
}

File: api/supabase-data-proxy.js
const DEFAULT_ALLOWED_RPCS = [
  'rpc_get_profile',
  'rpc_upsert_profile',
  'rpc_list_products',
  'rpc_get_product_variant_stocks',
  'rpc_list_best_seller_product_ids',
  'rpc_get_active_event',
  'rpc_get_wishlist_ids',
  'rpc_toggle_wishlist',
  'rpc_get_cart_items',
  'rpc_cart_total',
  'rpc_add_to_cart',
  'rpc_set_cart_qty',
  'rpc_clear_cart',
  'rpc_place_order',
  'rpc_get_orders',
  'rpc_get_order_support_statuses',
  'rpc_validate_promo_code',
  'rpc_upsert_payway_transaction',
  'rpc_app_log',
  'rpc_admin_list_profiles',
  'rpc_admin_set_account_type',
  'rpc_admin_list_orders',
  'rpc_staff_update_order_status',
  'rpc_staff_update_order_address',
  'rpc_staff_update_support_request_status',
  'rpc_staff_reply_product_rating',
  'rpc_admin_confirm_cash_payment',
  'rpc_staff_list_support_requests',
  'rpc_admin_list_events',
  'rpc_admin_create_event',
  'rpc_admin_update_event',
  'rpc_admin_delete_event',
  'rpc_admin_get_product_variant_stocks',
  'rpc_admin_set_product_variant_stocks',
  'rpc_admin_create_product',
  'rpc_admin_update_product',
  'rpc_admin_delete_product',
];
const {
  createRequestContext,
  durationMs,
  listObjectKeys,
  logError,
  logInfo,
  logWarning,
  safeError,
  sanitizeValue,
  truncate,
} = require('./_logging');

const DEFAULT_ALLOWED_TABLES = [
  'profiles',
  'events',
  'orders',
  'product_ratings',
  'payway_transactions',
];

function json(res, status, payload) {
  res.status(status).json(payload);
}

function readBody(req) {
  if (typeof req.body === 'string') {
    const value = req.body.trim();
    if (value.length === 0) return {};
    try {
      return JSON.parse(value);
    } catch (_) {
      return {};
    }
  }
  if (req.body && typeof req.body === 'object') {
    return req.body;
  }
  return {};
}

function parseAllowList(raw, defaults) {
  const value = (raw || '').trim();
  if (!value) return new Set(defaults);
  return new Set(
    value
      .split(',')
      .map((item) => item.trim())
      .filter((item) => item.length > 0),
  );
}

function encodeScalar(value) {
  if (value === null || value === undefined) return 'null';
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  return String(value);
}

function encodeInValue(value) {
  if (value === null || value === undefined) return 'null';
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value);
  }
  const text = String(value).replace(/"/g, '\\"');
  return `"${text}"`;
}

function encodeFilterExpression(op, value) {
  switch ((op || '').toString().trim().toLowerCase()) {
    case 'eq':
      return `eq.${encodeScalar(value)}`;
    case 'neq':
      return `neq.${encodeScalar(value)}`;
    case 'gt':
      return `gt.${encodeScalar(value)}`;
    case 'gte':
      return `gte.${encodeScalar(value)}`;
    case 'lt':
      return `lt.${encodeScalar(value)}`;
    case 'lte':
      return `lte.${encodeScalar(value)}`;
    case 'like':
      return `like.${encodeScalar(value)}`;
    case 'ilike':
      return `ilike.${encodeScalar(value)}`;
    case 'is':
      return `is.${encodeScalar(value)}`;
    case 'in': {
      const values = Array.isArray(value) ? value : [value];
      const encoded = values.map((item) => encodeInValue(item)).join(',');
      return `in.(${encoded})`;
    }
    default:
      return null;
  }
}

function buildOrderClause(orders) {
  if (!Array.isArray(orders) || orders.length === 0) return '';
  return orders
    .map((raw) => {
      const item = raw && typeof raw === 'object' ? raw : {};
      const column = String(item.column || '').trim();
      if (!column) return '';
      const ascending = item.ascending !== false;
      const nullsPart =
        typeof item.nullsFirst === 'boolean'
          ? item.nullsFirst
            ? '.nullsfirst'
            : '.nullslast'
          : '';
      return `${column}.${ascending ? 'asc' : 'desc'}${nullsPart}`;
    })
    .filter(Boolean)
    .join(',');
}

function summarizeFilters(filters) {
  if (!Array.isArray(filters)) return [];
  return filters.slice(0, 10).map((rawFilter) => {
    const filter =
      rawFilter && typeof rawFilter === 'object' ? rawFilter : {};
    return {
      column: String(filter.column || '').trim(),
      op: String(filter.op || '').trim().toLowerCase(),
      value: sanitizeValue(filter.value, String(filter.column || '')),
    };
  });
}

function summarizeOrders(orders) {
  if (!Array.isArray(orders)) return [];
  return orders.slice(0, 10).map((rawOrder) => {
    const order =
      rawOrder && typeof rawOrder === 'object' ? rawOrder : {};
    return {
      column: String(order.column || '').trim(),
      ascending: order.ascending !== false,
      nullsFirst:
        typeof order.nullsFirst === 'boolean' ? order.nullsFirst : undefined,
    };
  });
}

function summarizeValues(values) {
  if (Array.isArray(values)) {
    return {
      type: 'array',
      length: values.length,
      sampleKeys: values.slice(0, 3).map((item) => listObjectKeys(item)),
    };
  }
  return {
    type: values && typeof values === 'object' ? 'object' : typeof values,
    keys: listObjectKeys(values),
  };
}

async function forwardToSupabase({
  req,
  supabaseUrl,
  supabaseAnonKey,
  path,
  method,
  searchParams = null,
  body = null,
  prefer = '',
}) {
  const url = new URL(`${supabaseUrl.replace(/\/+$/, '')}${path}`);
  if (searchParams instanceof URLSearchParams) {
    searchParams.forEach((value, key) => url.searchParams.append(key, value));
  }

  const clientAuth =
    typeof req.headers.authorization === 'string'
      ? req.headers.authorization.trim()
      : '';
  const authorization = clientAuth || `Bearer ${supabaseAnonKey}`;

  const headers = {
    apikey: supabaseAnonKey,
    Authorization: authorization,
  };
  if (body !== null) {
    headers['Content-Type'] = 'application/json';
  }
  if (prefer) {
    headers['Prefer'] = prefer;
  }

  const response = await fetch(url, {
    method,
    headers,
    body: body === null ? undefined : JSON.stringify(body),
  });
  return response;
}

module.exports = async (req, res) => {
  const context = createRequestContext(req, 'supabase_data_proxy');
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader(
    'Access-Control-Allow-Headers',
    'Content-Type, Authorization',
  );

  logInfo('proxy.data.request_received', {
    requestId: context.requestId,
    scope: context.scope,
    method: context.method,
    path: context.path,
    clientIp: context.clientIp,
    hasAuthorizationHeader: context.hasAuthorizationHeader,
    userAgent: context.userAgent,
  });

  if (req.method === 'OPTIONS') {
    logInfo('proxy.data.preflight', {
      requestId: context.requestId,
      durationMs: durationMs(context),
    });
    res.status(204).end();
    return;
  }

  if (req.method !== 'POST') {
    logWarning('proxy.data.method_not_allowed', {
      requestId: context.requestId,
      method: req.method,
      durationMs: durationMs(context),
    });
    json(res, 405, {
      code: 'METHOD_NOT_ALLOWED',
      message: 'Use POST',
    });
    return;
  }

  const supabaseUrl = (process.env.SUPABASE_URL || '').trim();
  const supabaseAnonKey = (process.env.SUPABASE_ANON_KEY || '').trim();
  if (!supabaseUrl || !supabaseAnonKey) {
    logError('proxy.data.missing_supabase_config', {
      requestId: context.requestId,
      hasSupabaseUrl: supabaseUrl.length > 0,
      hasSupabaseAnonKey: supabaseAnonKey.length > 0,
      durationMs: durationMs(context),
    });
    json(res, 500, {
      code: 'MISSING_SUPABASE_CONFIG',
      message: 'SUPABASE_URL and SUPABASE_ANON_KEY are required',
    });
    return;
  }

  const allowedRpcs = parseAllowList(
    process.env.PROXY_ALLOWED_RPCS,
    DEFAULT_ALLOWED_RPCS,
  );
  const allowedTables = parseAllowList(
    process.env.PROXY_ALLOWED_TABLES,
    DEFAULT_ALLOWED_TABLES,
  );

  const body = readBody(req);
  const kind = String(body.kind || '').trim().toLowerCase();
  let operationSummary = {
    kind,
    bodyKeys: listObjectKeys(body),
  };

  let upstreamResponse;
  try {
    if (kind === 'rpc') {
      const rpc = String(body.rpc || '').trim();
      if (!rpc) {
        logWarning('proxy.data.invalid_rpc_name', {
          requestId: context.requestId,
          bodyKeys: listObjectKeys(body),
          durationMs: durationMs(context),
        });
        json(res, 400, {
          code: 'INVALID_RPC_NAME',
          message: 'rpc is required',
        });
        return;
      }
      if (!allowedRpcs.has(rpc)) {
        logWarning('proxy.data.rpc_not_allowed', {
          requestId: context.requestId,
          rpc,
          durationMs: durationMs(context),
        });
        json(res, 403, {
          code: 'RPC_NOT_ALLOWED',
          message: `RPC "${rpc}" is not allowed`,
        });
        return;
      }
      const params =
        body.params && typeof body.params === 'object' ? body.params : {};
      operationSummary = {
        kind,
        rpc,
        paramKeys: listObjectKeys(params),
      };
      logInfo('proxy.data.forwarding_rpc', {
        requestId: context.requestId,
        ...operationSummary,
      });
      upstreamResponse = await forwardToSupabase({
        req,
        supabaseUrl,
        supabaseAnonKey,
        path: `/rest/v1/rpc/${encodeURIComponent(rpc)}`,
        method: 'POST',
        body: params,
      });
    } else if (kind === 'table') {
      const action = String(body.action || '').trim().toLowerCase();
      const table = String(body.table || '').trim();
      if (!table) {
        logWarning('proxy.data.invalid_table_name', {
          requestId: context.requestId,
          bodyKeys: listObjectKeys(body),
          durationMs: durationMs(context),
        });
        json(res, 400, {
          code: 'INVALID_TABLE_NAME',
          message: 'table is required',
        });
        return;
      }
      if (!allowedTables.has(table)) {
        logWarning('proxy.data.table_not_allowed', {
          requestId: context.requestId,
          table,
          action,
          durationMs: durationMs(context),
        });
        json(res, 403, {
          code: 'TABLE_NOT_ALLOWED',
          message: `Table "${table}" is not allowed`,
        });
        return;
      }

      const filters = Array.isArray(body.filters) ? body.filters : [];
      const filterSummary = summarizeFilters(filters);
      const search = new URLSearchParams();
      for (const rawFilter of filters) {
        const filter =
          rawFilter && typeof rawFilter === 'object' ? rawFilter : {};
        const column = String(filter.column || '').trim();
        if (!column) continue;
        const expr = encodeFilterExpression(filter.op, filter.value);
        if (!expr) {
          logWarning('proxy.data.unsupported_filter_op', {
            requestId: context.requestId,
            table,
            action,
            column,
            op: String(filter.op || ''),
            durationMs: durationMs(context),
          });
          json(res, 400, {
            code: 'UNSUPPORTED_FILTER_OP',
            message: `Unsupported filter op "${filter.op}"`,
          });
          return;
        }
        search.append(column, expr);
      }

      if (action === 'select') {
        const columns = String(body.columns || '*').trim() || '*';
        search.set('select', columns);
        const orderClause = buildOrderClause(body.orders);
        if (orderClause) {
          search.set('order', orderClause);
        }
        if (Number.isInteger(body.limit) && body.limit > 0) {
          search.set('limit', String(body.limit));
        }
        operationSummary = {
          kind,
          table,
          action,
          columns,
          limit: Number.isInteger(body.limit) && body.limit > 0
            ? body.limit
            : undefined,
          filters: filterSummary,
          orders: summarizeOrders(body.orders),
        };
        logInfo('proxy.data.forwarding_table_select', {
          requestId: context.requestId,
          ...operationSummary,
        });
        upstreamResponse = await forwardToSupabase({
          req,
          supabaseUrl,
          supabaseAnonKey,
          path: `/rest/v1/${encodeURIComponent(table)}`,
          method: 'GET',
          searchParams: search,
        });
      } else if (action === 'update') {
        const values =
          body.values && typeof body.values === 'object' ? body.values : {};
        const returning = body.returning === true;
        const columns = String(body.columns || '*').trim() || '*';
        if (returning) {
          search.set('select', columns);
        }
        operationSummary = {
          kind,
          table,
          action,
          returning,
          columns: returning ? columns : undefined,
          filters: filterSummary,
          values: summarizeValues(values),
        };
        logInfo('proxy.data.forwarding_table_update', {
          requestId: context.requestId,
          ...operationSummary,
        });
        upstreamResponse = await forwardToSupabase({
          req,
          supabaseUrl,
          supabaseAnonKey,
          path: `/rest/v1/${encodeURIComponent(table)}`,
          method: 'PATCH',
          searchParams: search,
          body: values,
          prefer: returning ? 'return=representation' : 'return=minimal',
        });
      } else if (action === 'upsert') {
        const values = body.values;
        const onConflict = String(body.onConflict || '').trim();
        const returning = body.returning === true;
        const columns = String(body.columns || '*').trim() || '*';
        if (onConflict) {
          search.set('on_conflict', onConflict);
        }
        if (returning) {
          search.set('select', columns);
        }
        const prefer = returning
          ? 'resolution=merge-duplicates,return=representation'
          : 'resolution=merge-duplicates,return=minimal';
        operationSummary = {
          kind,
          table,
          action,
          onConflict: onConflict || undefined,
          returning,
          columns: returning ? columns : undefined,
          values: summarizeValues(values),
        };
        logInfo('proxy.data.forwarding_table_upsert', {
          requestId: context.requestId,
          ...operationSummary,
        });
        upstreamResponse = await forwardToSupabase({
          req,
          supabaseUrl,
          supabaseAnonKey,
          path: `/rest/v1/${encodeURIComponent(table)}`,
          method: 'POST',
          searchParams: search,
          body: values,
          prefer,
        });
      } else {
        logWarning('proxy.data.invalid_table_action', {
          requestId: context.requestId,
          table,
          action,
          durationMs: durationMs(context),
        });
        json(res, 400, {
          code: 'INVALID_TABLE_ACTION',
          message: 'action must be one of: select, update, upsert',
        });
        return;
      }
    } else {
      logWarning('proxy.data.invalid_proxy_kind', {
        requestId: context.requestId,
        kind,
        bodyKeys: listObjectKeys(body),
        durationMs: durationMs(context),
      });
      json(res, 400, {
        code: 'INVALID_PROXY_KIND',
        message: 'kind must be "rpc" or "table"',
      });
      return;
    }
  } catch (error) {
    logError('proxy.data.upstream_request_failed', {
      requestId: context.requestId,
      durationMs: durationMs(context),
      operation: operationSummary,
      error: safeError(error),
    });
    json(res, 502, {
      code: 'UPSTREAM_REQUEST_FAILED',
      message: error instanceof Error ? error.message : 'Unknown proxy error',
    });
    return;
  }

  const contentType =
    upstreamResponse.headers.get('content-type') ||
    'application/json; charset=utf-8';
  const rawText = await upstreamResponse.text();
  const responseBody = rawText.trim().length === 0 ? '{}' : rawText;
  if (upstreamResponse.status >= 400) {
    logWarning('proxy.data.upstream_failed', {
      requestId: context.requestId,
      durationMs: durationMs(context),
      operation: operationSummary,
      upstreamStatus: upstreamResponse.status,
      contentType,
      responsePreview: truncate(responseBody, 240),
    });
  } else {
    logInfo('proxy.data.upstream_succeeded', {
      requestId: context.requestId,
      durationMs: durationMs(context),
      operation: operationSummary,
      upstreamStatus: upstreamResponse.status,
      contentType,
    });
  }
  res.status(upstreamResponse.status);
  res.setHeader('Content-Type', contentType);
  res.send(responseBody);
};
